app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -5,8 +5,30 @@ class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
   skip_after_action :verify_policy_scoped
 
   def cis2
-    @user =
-      User.find_or_create_user_from_cis2_oidc(request.env["omniauth.auth"])
-    sign_in_and_redirect @user, event: :authentication # this will throw if @user is not activated
+    user_info = request.env["omniauth.auth"]
+    raw_info = user_info["extra"]["raw_info"]
+    selected_role = raw_info["selected_roleid"]
+    nrbac_role =
+      raw_info["nhsid_nrbac_roles"].find do
+        _1["person_roleid"] == selected_role
+      end
```

Comment:
I feel like the controller action would be skinnier if we extracted this assignment logic, and perhaps the flash-setting bits further down, to private methods... It would introduce some indirection, but would make it easier to see at a high level what the action is doing.

Reply:
Yup, will tidy up.

Reply:
I've done a bit of work here including adding a concern and closing up another edge-case where someone could visit pages without having their team signed up (albeit without seeing any data). Please re-review.